node_modules/oae-principals/lib/definitiveDeletion.js

```diff
@@ -0,0 +1,1672 @@
+                        if (err) {
+                            return callback(err);
+                        }
+
+                        // Add element to archive
+                        _addToArchive(archiveUsers.archiveId, user.id, listIdElement, function(err) {
+                            if(err) {
+                                return callback(err);
+                            }
+
+                            // Delete followers and following
+                            _deleteFollowers(ctx, user, function(err) {
+                                if (err) {
+                                    return callback(err);
+                                }
+                                _deleteFollowing(ctx, user, function(err) {
+                                    if (err) {
+                                        return callback(err);
+                                    }
+
+                                    // Send all emails
+                                    _sendEmail(ctx, listEmail, function(err, listEmail) {
+                                        if (err) {
+                                            return callback(err);
+                                        }
+                                        
+                                        return callback(null, listEmail);
+                                    });
+                                });
+                            });
+                        });
+                    });
+                });
+            });
+        });
+    });            
+};
+
+/**
+ * Send an email token to a user that have a resource shared with the deleted user
+ *
+ * @param  {data}       data            List of the user to send the email token to with the resource concerned
+ * @param  {data}       idUserDeleted   Id of the user deleted
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Object}     callback.err    An error that occurred, if any
+ */
+ // TODO(claire)
+var _parseData = function(data, idUserDeleted, callback) {
+
+    // values of deleted :
+    //      userWillBePromoted : list of user (and resource concerned) who will recieve an email 'some user will be promoted'
+    //      contentWillBeDeleted : list of user (and resource concerned) who will recieve an email 'resource will be removed'
+    //      userJustLeaving : list of user (and resource concerned) who will recieve an email 'this user will leave the resource'
+
+    if (_.isEmpty(data[0].memberList)) {
+        return callback();
+    }
+    var listElementByMember = {'users': [] };
+
+    // Fill the list with a first element
+    listElementByMember.users.push({'userWillBePromoted': [], 'contentWillBeDeleted': [], 'userJustLeaving': [], 'profile': data[0].memberList[0].profile});
+    
+    var exist = false;
+
+    // For each member/editor/... of a resource
+    for (var i=0; i<=data.length; i++) { 
```

Comment:
After going over the other PR I thought I'd point out that a similar structure is used here as well and that generally such deep nesting of for-loops should be avoided.